### ISSUE TITLE
Duplicate checker for Article Post

### ISSUE THREAD

**Issue by madzadev** [forem/forem#7997]
**Is your feature request related to a problem? Please describe.**

DEV.to really is an awesome platform and it's like a morning coffee to me. Tho yesterday I came across this unpleasant experience.

I was working for past couple of days on the article, published it and 5 mins after I saw other user had copied everything word for word and created it as his article.

Original: https://dev.to/madarsbiss/how-it-feels-like-to-style-things-in-2020-21cm
Copy-paste: https://dev.to/deepanshu1422/how-it-feels-like-to-style-things-in-2020-4p0e

**Describe the solution you'd like**

I you want to stand by your writers, keep their interest of coming back and increase the overall uniqueness and quality of content (and that's why readers are there), a good place to start would be a duplicate checker:

1.User opens new article modal
2.Write's his/her article
3.Click Post
4.Then content get's evaluated in back-end, comparing it to articles in DEV.to database or even better - to DEV.to database and outside sources. U can just show a modal for user "Content is being processed" or something during this.
5.Finally, if everything passes, article gets published, if not show modal to user on what should be improved.

**Describe alternatives you've considered**

**Additional context**

I am considering this a main issue. 
*Please forward this to content managers and remove the copy, btw.

**Comment by citizen428** [comment]
Thanks for the issue! We'll take your request into consideration and follow up if we decide to tackle this issue.

To our amazing contributors: issues labeled `type: bug` are always up for grabs, but for feature requests, please wait until we add a `ready for dev` before starting to work on it.

To claim an issue to work on, please leave a comment. If you've claimed the issue and need help, please ping @thepracticaldev/oss and we will follow up within 3 business days.

For full info on how to contribute, please check out our [contributors guide](https://github.com/thepracticaldev/dev.to/blob/master/CONTRIBUTING.md).

**Comment by citizen428** [comment]
I'm sorry this happened @madarsbiss. Please note that you can report this to our support team:

<img width="424" alt="How_it_feels_like_to_style_things_in_2020__-_DEV_Community_👩💻👨💻" src="https://user-images.githubusercontent.com/47985/82546238-0333c680-9b82-11ea-8fac-a50304878b44.png">

I've already reported this specific issue for you though.

**Comment by madarsbiss** [comment]
@citizen428 didn't know there was a report abuse option and thanks for reporting.. It would still be nice to implement this feature to reduce cases like these in the future! really hope DEV team will decide to tackle it :)

**Comment by maniflames** [comment]
Reported the article as well! 
I'd love to see a duplicate checker though, the idea that your work can just be copy-pasted and posted on the exact same platform feels really discouraging. 

I tried to comment on it but the author just decide to hide my comment :/

**Comment by madarsbiss** [comment]
@maniflames thanks! united we stand!

**Comment by citizen428** [comment]
> I'd love to see a duplicate checker though, the idea that your work can just be copy-pasted and posted on the exact same platform feels really discouraging.

Out of interest (and for reference), how do other content sites like e.g. Medium handle this?

**Comment by maniflames** [comment]
I'm not on medium at all because I really don't like the experience. However I'm still a student and if there is an 80% match between what I and a classmate have written it is considered plagiarism and we'll have to go through the exams committee to sort things out. There is an existing database of all the work that has been handed in before and some popular sources on the internet. Unfortunately I don't know the exact method that is used.

I've googled a bit and right now Medium does the same thing as DEV and responds to DMCA claims [(source)](https://help.medium.com/hc/en-us/articles/360041640213-Plagiarism-Guidelines).

I understand that plagiarism is a hard nut to crack but it's very unlikely that a new post of 1500+ words that matches 98% (copycat put some phrases in bold and italics) with an existing post is an accident. I'll try to figure out how plagiarism is detected by software that schools use (or at least look for methods that resemble it). I hope this helps a little bit.

**Comment by citizen428** [comment]
Thanks for doing this @maniflames, much appreciated!

**Comment by maniflames** [comment]
I didn't really expect this but it turns out there is an entire field dedicated to [content/text similarity](https://en.wikipedia.org/wiki/Content_similarity_detection)😅. There are different types of plagiarism that you could put on a spectrum. Different methods are successful at detecting different types of plagiarism.

![https://upload.wikimedia.org/wikipedia/commons/thumb/6/67/PD_Methods_Detection_Performance.png/2880px-PD_Methods_Detection_Performance.png](https://upload.wikimedia.org/wikipedia/commons/thumb/6/67/PD_Methods_Detection_Performance.png/2880px-PD_Methods_Detection_Performance.png)
(Image from the [content similarity](https://en.wikipedia.org/wiki/Content_similarity_detection) wikipedia page)

For something as obvious as in this case (literal copy paste) looking for matching substrings or some version of 'git diff' and processing the results should be enough. To detect something something that hidden a little better you might want to use the methods that schools use.

The best example I could find of software that compares submissions and returns pairs of possible matches at scale is TAPS (Tool of Addressing Plagiarism at Scale). It's build upon Stanford's MOSS (Measure Of Software Similarity) and used to check submissions of code at a Udacity course. There is a [YouTube video](https://youtu.be/pFNA8RKt79Q) with a link towards [a paper](http://lucylabs.gatech.edu/b/wp-content/uploads/2016/04/wip146-sheahenA.pdf) in the description. 

To summarise how TAPS & MOSS work together:

- Input new & past project submissions
- Whitelist boilerplate code & expected solutions
- Streamline input for MOSS
(Everything below is how MOSS works internally, this [blogpost](https://yangdanny97.github.io/blog/2019/05/03/MOSS) explains the algorithm in more detail)
- For every input file do the following comparison against all other input files:
    - Remove irrelevant features like whitespace from text
    - Generate fingerprints using the winnowing algorithm
    - Compare fingerprints
- Only return reports of matching submissions that contain a new project

There is definitely room for performance improvement that shows in the general TAPS algorithm in DEV's case but since this is a utility build specifically for code we can only use the general idea anyway (it's optimised for specific programming languages). I've tried to figure out how popular educational checkers like [Unicheck](https://unicheck.com/plagiarism-checker-features), [Turnitin](https://www.turnitin.com/) & [Copyleaks](https://api.copyleaks.com/) work but unfortunately all I could find was 'scans against X Billion web pages & academic papers' (external sites should probably be out of scope for DEV). There are also specific ML/NLP algorithms that can check how similar words are so plagiarism that is hidden by replacing words with synonyms can be detected too. It is most likely best to no go overboard and start with the least complicated method that targets the desired level of plagiarism. 

I did a quick search for relevant Ruby Gems but most seem kind of abandoned. The APIs I found so far are paid and check specifically against webpages. I hope this help a little more, I can do a more in dept search of existing solutions or help out with designing a custom solution for DEV in case help is needed. (I am way more into this that thought 😂)

**Comment by citizen428** [comment]
@maniflames Wow, thank you so much for doing all of this! Before you do any more research on this, keep in mind that this feature has not been approved yet. I wouldn't want you to sink too much time into this and then be disappointed if we decide not to tackle this specific feature. Unless you're super into the topic now, then please do keep us posted 😄

**Comment by cmgorton** [comment]
@michael-tharrington could you weigh in on the issue at all with how we currently handle duplicate articles. I am pretty sure right now we rely heavily on our trusted users to report signs of plagiarism right?

**Comment by michael-tharrington** [comment]
That's right! We rely on reports from the community via https://dev.to/report-abuse and trusted users who help keep an eye out for conduct-breaking material. That said, I LOVE the idea of us having a duplicate content checker. Often when someone reports plagiarism, I navigate to plagium.com and run chunks of text through the free search option they have, but it's time consuming.

Side note — I read this post previously  - https://dev.to/jessesbyers/someone-stole-my-dev-article-how-to-build-a-python-script-to-detect-stolen-content-54fg and reached out to the DEV user there to see if they might be interested in creating something like this for DEV. Not sure if it's helpful, but figured I'd give it a mention.

**Comment by cmgorton** [comment]
Thanks Michael! I am going to label this as a potential RFC so the team can discuss this feature more.

**Comment by cmgorton** [comment]
Since this issue has gone stale I am going to close it. If anyone would like to still implement this feature let me know and we can turn this issue into a [discussion](https://github.com/forem/forem/discussions)
